Sources/Algorithms/MinMax.swift

```diff
@@ -383,3 +383,92 @@ extension Collection where Element: Comparable {
     return max(count: count, sortedBy: <)
   }
 }
+
+//===----------------------------------------------------------------------===//
+// Simultaneous minimum and maximum evaluation
+//===----------------------------------------------------------------------===//
+
+extension Sequence {
+  /// Returns both the minimum and maximum elements in the sequence, using the
+  /// given predicate as the comparison between elements.
+  ///
+  /// The predicate must be a *strict weak ordering* over the elements. That
+  /// is, for any elements `a`, `b`, and `c`, the following conditions must
+  /// hold:
+  ///
+  /// - `areInIncreasingOrder(a, a)` is always `false`. (Irreflexivity)
+  /// - If `areInIncreasingOrder(a, b)` and `areInIncreasingOrder(b, c)` are
+  ///   both `true`, then `areInIncreasingOrder(a, c)` is also
+  ///   `true`. (Transitive comparability)
+  /// - Two elements are *incomparable* if neither is ordered before the other
+  ///   according to the predicate. If `a` and `b` are incomparable, and `b`
+  ///   and `c` are incomparable, then `a` and `c` are also incomparable.
+  ///   (Transitive incomparability)
+  ///
+  /// This example shows how to use the `extrema(by:)` method on a
+  /// dictionary to find the key-value pair with the lowest value and the pair
+  /// with the highest value.
+  ///
+  ///     let hues = ["Heliotrope": 296, "Coral": 16, "Aquamarine": 156]
+  ///     let minmax = hues.extrema { a, b in a.value < b.value }
+  ///     let leastHue = minmax?.min, greatestHue = minmax?.max
+  ///     print(leastHue, greatestHue)
+  ///     /*
+  ///       Prints:
+  ///         """
+  ///         Optional((key: "Coral", value: 16)) Optional((key: "Heliotrope", value: 296))
+  ///         """
+  ///      */
+  ///
+  /// - Parameter areInIncreasingOrder: A predicate that returns `true`
+  ///   if its first argument should be ordered before its second
+  ///   argument; otherwise, `false`.
+  /// - Returns: A tuple with the sequence's minimum element, followed by its
+  ///   maximum element. For either member, if the sequence provides multiple
+  ///   qualifying elements, the one chosen is unspecified. The same element may
```

Comment:
I’d like to take a more principled stand than this. Right now we have two different behaviors for which maximum element is chosen. First, `sort()`, `min(count:)`, and `Swift.max` (not the sequence version) all either return the last equal value or, in the sort/min(count:) case, put it at the end. 

```swift
/// All instances of `Foo` are equal to each other.
class Foo: Comparable {
    static func ==(lhs: Foo, rhs: Foo) -> Bool { true }
    static func < (lhs: Foo, rhs: Foo) -> Bool { false }
}

let a = Foo()
let b = Foo()
max(a, b) === b            // true
[a, b].sorted().last === b // true
```

Second, the sequence version of `max()` returns the _first_ instance of the maximum value.

```swift
[a, b].max() === b         // false
```

On the `min` side, there’s only one behavior — all the equivalent methods return the first minimum element.

Even though we’ll end up where this method and the stdlib’s `Sequence.max()` return different instances, let’s align this new method with the majoritarian position, and I’ll keep looking at how we can switch `Sequence.max()` to come into alignment.

Reply:
Enshrining which element is chosen to break ties seems like an over-specification; someone's code may be buggy if they're depending on a specific equivalent element.

Reply:
We already specify this in e.g. `Swift.max`, and I'd like to make the `sort()` stability a guarantee, which is similar. The dependence is exactly why I'd like to specify the behavior — rather than someone depending on the result of an implementation quirk, they can depend on documented behavior, and we can test to make sure we meet the expectation.